File: 350. Intersection of Two Arrays II.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dic1 = {}
        for i in nums1:
            dic1[i] = dic1.get(i, 0) + 1
        dic2 = {}
        for i in nums2:
            dic2[i] = dic2.get(i, 0) + 1
        
        if len(dic1) > len(dic2):
            dic1, dic2 = dic2, dic1
        
        res = []
        for i in dic1:
            if i in dic2:
                res += [i] * min(dic1[i], dic2[i])
                
        return res

File: 350. Intersection of Two Arrays II.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i : nums1) {
            map.put(i, map.getOrDefault(i, 0)+1);
        }
        List<Integer> ls = new ArrayList<>();
        for (int i : nums2) {
            if (map.getOrDefault(i, 0) > 0) {
                ls.add(i);
                map.put(i, map.get(i)-1);
            }
        }
        int[] ans = new int[ls.size()];
        int index = 0;
        for (int i : ls) {
            ans[index++] = i;
        }
        return ans;
    }
}